### ISSUE TITLE
Watch - Doesn't update

### ISSUE THREAD

**Issue by bluechips23** [kubernetes-client/java#279]
I have two tasks that I run in parallel threads and I'm pulling my hair on why the `Watch` functionality doesn't work. Please let me know if you have any insights.

**Task 1:** Get the status of the pods and show the current status.
**Task 2:** Keep a watch on new events. This is what I'm trying to understand better.

Each of these tasks are executed every 30 seconds with `scheduledAtFixedRate()`.

**Expected behavior:** 
**Task 1:** I should get the list of all the pods with their current status (this works). 
**Task 2:** I should expect list of new events as they happen.

**Observed behavior:** 
**Task 1:** It works fine. I get updated status of the pods every 30 seconds.
**Task 2:** It dumps the events from the first request, but it seems like it doesn't update any new events.

**Code:**
**Task 1:**

```
@Component
@Scope(value = org.springframework.beans.factory.config.ConfigurableBeanFactory.SCOPE_SINGLETON)
public class Task1 implements Runnable {

    private ScheduledExecutorService scheduledExecutorService;
    private CommandInvoker commandInvoker;
    private static final int INITIAL_DELAY = 15; 
    private static final int POLLING_INTERVAL = 30;

    @Autowired
    public Task1 (CommandInvoker commandInvoker,
                        ScheduledExecutorService scheduledExecutorService) {
        this.commandInvoker = commandInvoker;
        this.scheduledExecutorService = scheduledExecutorService;
        this.scheduledExecutorService.scheduleAtFixedRate(this, INITIAL_DELAY, POLLING_INTERVAL, TimeUnit.SECONDS);
    }

    @Override
    public void run() {
        System.out.println("===== STARTING TASK 1 POD HEALTH CHECK =======");
        commandInvoker.getPodStatus();
    }
}
```

**Task 2:**

```
@Component
@Scope(value = org.springframework.beans.factory.config.ConfigurableBeanFactory.SCOPE_SINGLETON)
public class Task2 implements Runnable {

    private ScheduledExecutorService scheduledExecutorService;
    private CommandInvoker commandInvoker;

    @Autowired
    public Task2(CommandInvoker commandInvoker,
                              ScheduledExecutorService scheduledExecutorService) {
        this.commandInvoker = commandInvoker;
        this.scheduledExecutorService = scheduledExecutorService;
        this.scheduledExecutorService.scheduleAtFixedRate(this, 30, 30, TimeUnit.SECONDS);
    }

    @Override
    public void run() {
        System.out.println("===== STARTING TASK 2 EVENT WATCH UPDATE =======");
        commandInvoker.getWatchUpdates();
    }
}
```

**CommandInvoker:**

```
@Component
public class CommandInvoker {

    public void getPodStatus() {
        try {
            CoreV1Api api = new CoreV1Api();
            V1PodList list = api.listPodForAllNamespaces(null,
                    null, null,  null, null, null, null, null, null);
            for( V1Pod pod : list.getItems() ) {
                  // THIS WORKS //
            }
        } catch ( ApiException e) {
            throw new WhateverException ("Failed to handle watchlist event", e);
        }
    }

    public void getWatchUpdates() {
        CoreV1Api api = new CoreV1Api();
        try {
            Watch<V1Event> watch = Watch.createWatch(
                    apiClient,
                    api.listEventForAllNamespacesCall(null, null, null, null,
                            null, null, null, null, true, null, null),
                    new TypeToken<Watch.Response<V1Event>>() {}.getType());

            watch.forEach( response  -> {
                V1Event event = response.object;
                // THIS ONLY DUMPS EVENTS FROM FIRST CALL BUT NEVER GETS EXECUTED AGAIN
            });
              // I NEVER REACH HERE BUT I DON'T GET ANY UPDATES
        } catch ( ApiException e) {
            throw new K8ServerException("Failed to handle watchlist event", e);
        }
    }
}
```

**Comment by brendandburns** [comment]
You are watching `Events` objects, but you are listing `Pod` objects.

I assume that you actually want to watch for changes to `Pod` objects, if that's the case, you need to watch `Pod` instead of `Event` (e.g. `listPodForAllNamespacesCall`)

Hope that helps.

**Comment by bluechips23** [comment]
> You are watching Events objects, but you are listing Pod objects.
> 
> I assume that you actually want to watch for changes to Pod objects, if that's the case, you need to watch Pod instead of Event (e.g. listPodForAllNamespacesCall)

In this example here, its true that I’m watching for Events, but honestly, I’m have also tried this with watching for Pods and Deployments and I’m seeing the same behavior. The common issue among all of them is that the Watch stops listing new changes in pods/events/deployments after the initial first request and when you have another task for listing the pods or listing the deployments. 

The only way I have seen Watch working is if I create a Watch (of anything) as a standalone task. But if you have a running Watch, and you create another task for listing the pods (or any other API call to K8s cluster), the Watch stops updating.

If you have some suggestions on what further tests to do, I’ll be happy to try them out.

**Comment by brendandburns** [comment]
Interesting. I wonder if there issue has to do with sharing an http client....

If you send me a simple repro I can try to debug in my side too...

Thanks!

--brendan

________________________________

**Comment by brendandburns** [comment]
@bluechips23 can you see this comment: https://github.com/kubernetes-client/java/issues/275#issuecomment-392022443

Is it possible you have debugging enabled too?

**Comment by brendandburns** [comment]
@bluechips23 any confirmation that this is due to Debugging being enabled?

Thanks!

**Comment by bluechips23** [comment]
Hi @brendandburns - my apologies in delayed response. I was on vacation last week without access to my code.

I just checked - I don't have any debugging enabled. I'm creating my ApiClient this way (there's more involved but this is the heart of it):

```
ApiClient apiClient = new ClientBuilder()
                    .setVerifyingSsl(false)
                    .build();
            apiClient.setBasePath(connection.getBaseUrl());
```

**Comment by brendandburns** [comment]
Ok I'll try to repro...

--brendan

________________________________

**Comment by bluechips23** [comment]
@brendanburns - I forked the client-java into my repo and I added a new example which reproduces the issue that I am having. Please take a look at https://github.com/bluechips23/java/blob/master/examples/src/main/java/io/kubernetes/client/examples/ComboWatchExample.java 

You should be able to check this out and reproduce the issue.

**Comment by brendandburns** [comment]
Cool I'll try to look soon.

--brendan

________________________________

**Comment by rjeberhard** [comment]
@bluechips23, I don't know if it will help you, but we have an operator based on this library that makes heavy use of watches.  We've built up some machinery around the API and watches to pool connections and to restart watches after errors or timeouts (preserving last seen resourceVersion, etc.). I've been meaning to offer some of this back to @brendandburns but haven't had the chance to separate it from our project.  However, if there is anything here that helps you, obviously have at it.  I can assert that we've run this watch code through all kinds of negative testing and aren't seeing the failures you describe.  Note: at the bottom we are using hasNext / next rather than forEach.  https://github.com/oracle/weblogic-kubernetes-operator/tree/develop/operator/src/main/java/oracle/kubernetes/operator/builders

**Comment by brendandburns** [comment]
@rjeberhard would love to see any improvements for Watch (or even docs for advice/best practices) as PR for this repo...

Thanks!

**Comment by adohe** [comment]
@brendandburns any thoughts on this issue? is it related with the sharing an http client?

**Comment by todd-newton-relativity** [comment]
Any updates on this ticket? I'm running into similar issues when I try to create a watch on Jobs. The watch never updates.
